Add Guest memory alerts

sradco opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
#### Before this PR:
There is no alert for cases where the VM is experiencing high memory pressure.
#### After this PR:

Adds two guest-OS memory alerts. Guest signals reflect true application pressure better than host/cgroup views.

- KubeVirtVMGuestMemoryPressure (warning, for: 5m)
  - Fires when: available/total < 5% AND (major faults > 5/s OR swap IO > 1 MiB/s).
  - Why these thresholds: 5% headroom + sustained >5 pgmajfaults/s or significant swap IO indicates real pressure/thrashing; 5m window filters brief bursts and alerts well before OOM.

- KubeVirtVMGuestMemoryAvailableLow (info, for: 30m)
  - Fires when: available/total < 3% for 30m AND negligible swap IO (<2 KiB/s) AND low major faults (<1/s).
  - Why these thresholds: Flags persistently tiny headroom in guests likely without swap; long window avoids noise and excludes thrashing cases already covered by the Pressure alert.

Defaults are conservative and can be tuned per workload/SLO.

Assisted-by: Cursor for creating Prometheus unit tests.

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes #
- Partially addresses #
-->
Fixes #https://issues.redhat.com/browse/CNV-54116

<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issue/<vep_tracking_issue_number>
-->

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
The KubeVirtVMGuestMemoryPressure
and KubeVirtVMGuestMemoryAvailableLow alerts were added to alert when a VM memory is in pressure.
```

sradco commented:
Hi @enp0s3 , I updated the alerts based on our discussion.
Please review.

avlitman commented:
need to add runbook and consider adding e2e tests

enp0s3 reviewed (commented):
@sradco Thank you for the effort! It looks like a good path forward. It would be great to add to the PR description the reasoning for the metrics that were chosen and their corresponding thresholds. 

For example: 
```
rate(kubevirt_vmi_memory_pgmajfault_total[5m])) > 5
```
1. Why the value 5 was chosen as threshold?
2. How do we know that the rate of 5 for 5 minutes won't crash the VM way before we alert?

Lets add descriptive information about each such query.

sradco commented:
/test pull-kubevirt-check-tests-for-flakes

avlitman commented on tests/monitoring/monitoring.go:
```diff
@@ -162,6 +165,48 @@ var _ = Describe("[sig-monitoring]Monitoring", Serial, decorators.SigMonitoring,
 		})
 	})
 
+	Context("VM Memory Alerts", func() {
```
@sradco both e2e tests added in this pr fail

avlitman commented on tests/monitoring/monitoring.go:
```diff
@@ -162,6 +165,48 @@ var _ = Describe("[sig-monitoring]Monitoring", Serial, decorators.SigMonitoring,
+			}, 30)).To(Succeed())
+
+			By("Waiting for the low available memory alert")
+			libmonitoring.VerifyAlertExistWithCustomTime(virtClient, "KubeVirtVMGuestMemoryAvailableLow", 5*time.Minute)
```
use `libmonitoring.VerifyAlertExist` like other tests it's calling `VerifyAlertExistWithCustomTime` with 120 seconds. 5 minutes is too much also not make sense since using `libmonitoring.ReduceAlertPendingTime(virtClient)`

avlitman commented:
/approve

enp0s3 reviewed (commented):
@sradco Thank you, overall looks good. A small nit regarding the PR description:
```
Assisted-by: Cursor for creating e2e tests.
```
I don't see e2e tests in the PR, did you mean the alert unit tests? if so can you please update the signature?

sradco committed:
Add Guest memory alerts

The PR adds the KubeVirtVMGuestMemoryPressure
and KubeVirtVMGuestMemoryAvailableLow alerts.

The alert uses guest OS perspective for accurate
resource monitoring.

Signed-off-by: Shirly Radco <sradco@redhat.com>

enp0s3 commented:
/lgtm
@sradco Thank you!
